oeps/oep-0041-arch-async-server-event-messaging.rst

```diff
@@ -0,0 +1,486 @@
+services. There has also been a strong desire for a properly documented set of
+integration APIs that would allow Open edX to better interoperate with other
+applications in the ecosystem.
+
+There is an opportunity to bring all these elements together and create a solid
+foundation for simple, reliable, well documented, near real-time cross-service
+message passing. But without some planning and design work, we may end up in an
+even more chaotic place where each pairing of services uses different
+conventions for this kind of communication.
+
+
+-------------
+Specification
+-------------
+
+Message Format
+==============
+
+The message format will use the `CloudEvents 1.0
+<https://github.com/cloudevents/spec/blob/master/spec.md>`_ specification, with
+additional usage conventions that all Open edX services should adhere to.
+CloudEvents was developed by the Cloud Native Computing Foundation, with
+participation from major companies like Google, Microsoft, IBM, Red Hat, Oracle,
+and Huawei. While the finalized 1.0 spec is only months old and the Python
+tooling is primitive (advertised to break at any time), CloudEvents is also a
+pretty minimal message envelope spec, with the basic required fields being
+things that we'd want to have anyway.
+
+Event messages will be UTF-8 encoded JSON documents of 64 KB or less. An
+example:
+
+.. code-block:: javascript
+
+  {
+    "id": "b3b3981e-7544-11ea-b663-acde48001122",
+    "source": "https://discovery.edx.org/events/"
+    "specversion": "1.0",
+    "type": "org.openedx.catalog.course.created.v1"
+    "time": "2020-02-23T09:43:18Z",
+    "minorversion": 0,
+    "data": {
+        // message specific payload here
+    }
+  }
+
+CloudEvents Fields
+==================
+
+All top level field names are specified by the CloudEvents spec, but most fields
+defined by that standard have some leeway in how they're used for any given
+system. Each field below has a link to CloudEvents requirements, as well as a
+description of the conventions Open edX will use for that field. All Open edX
+events should have all of the fields specified below (even if some are
+considered optional in the CloudEvents spec).
+
+`id <https://github.com/cloudevents/spec/blob/master/spec.md#id>`_
+------------------------------------------------------------------
+Example: ``"b3b3981e-7544-11ea-b663-acde48001122"``
+
+Message ID. CloudEvents requires this to be a string that is unique for any
+given source. We will use a UUID1 here, since it is easy to generate without
+worrying about collisions, and it offers slightly more useful introspection
+(timestamp and machine) than a random UUID4. The UUID will be encoded as a
+string using Python's default behavior: lowercase and and dash-separated.
+
+`source <https://github.com/cloudevents/spec/blob/master/spec.md#source-1>`_
+-----------------------------------------------------------------------------
+Example: ``"https://discovery.edx.org/events/"``
```

Comment:
The source identifier should not change over time. Yet, it is quite possible to switch an application from http to https, and vice versa, based on just a configuration value, so I do not recommend including the scheme in the source name.

Also, should we use the external or the internal domain name? The edX discovery service might be available at discovery.edx.org, but it might also be available at discovery1.edx.org. If the discovery service runs inside a docker-compose or Kubernetes cluster, its internal domain name will simply be "discovery".

Thus, I suggest  we do not rely on http/https urls as the source name. Instead, I propose an identifier similar to the "application-specific identifiers" described in the spec: https://github.com/cloudevents/spec/blob/master/spec.md#source-1

Something along the lines of: "/openedx/discovery/events".

Reply:
Thank you for bringing this up! This was quite possibly the hardest field for me to reason about, and I'm grateful to get feedback on it. I'm not really happy with what I've got written here, but I'm not sure about the implications of something like "/openedx/discovery/events" either.

I originally had an app specific hierarchy like you propose, but I was later swayed by the CloudEvents spec recommendation that it be an absolute URI (which I realize is a recommendation and not a requirement). I'm definitely not tied to the current representation, but I do want to try to figure out exactly what it is we want to accomplish with this field. The only thing the spec really requires is that this value is a string and that `source + id` is unique (which we get for free anyway since we use UUIDs), so we have a lot of freedom in how we want to interpret this field.

One goal could be for the source to represent a sort of top level schema, e.g. "This is a family of events associated with a subset of Open edX." I believe that's what they're going for with their `https://github.com/cloudevents` example. I feel like we could accomplish this with `dataschema` if we wanted to. Another possible goal would be to map to something more concrete and physical with respect to identifying the producers. The spec also gives the example of having a unique identifier for every distinct producer.

We already encode the logical hierarchy in the message type, e.g. `org.openedx.catalog.course.created`. I felt that it would be redundant to have the source represent that concept. So I wanted to focus on the issue of, "Where did this event originate?" for debugging purposes.

> The source identifier should not change over time.

I think there are at least three use cases where the "source" service might change over time:

1. Service is renamed, e.g. course-discovery to catalog.
2. A service is refactored or broken up into multiple services that still share responsibility for some domain. Downstream processes listening for events still just listen for the event type they're interested in and don't notice any changes.
3. An Open edX service is entirely replaced by another system that emits the same events to make it easier to maintain compatibility. If the LMS is going to listen for course run `start_date` changes coming from the catalog, you could imagine something like Richie taking over that role of emitting those events for some organizations.

So for those use cases, I'm actually assuming that the source for a given message can change, and that it's a bad smell if any of our code relies on it being stable or on introspecting it. I went with absolute URIs because I thought this would simplify the namespacing and make it easier to debug issues. But I like your suggestion better and I think it could get us all the things that I really want from debugging, if we did something like:

source = "/{namespace}/{app}/{worker|web}"

So in the current scheme of things:
* "/openedx/discovery/web" for something coming from Django web requests to the course discovery service.
* "/openedx/discovery/worker" if the event is coming from one of its celery workers.

If we're logging that, it would be easy to track down where the message is coming from, and especially if we have multiple producers of the event. It would also have the added benefit that if something popular-but-beyond-our-control emits events with a weird data quirk, we could find and "fix" those events by inspecting their source. Though that should be avoided whenever possible.

A downside is that the whole "aid debugging" goal completely fails if people start relying on the source and producers have to start lying to not break things...

Does that sound good to you?

Reply:
Sounds very good! :ok_hand: Thanks for the detailed answer :)

Reply:
I like the new format you're thinking about @ormsbee , do you think source sort of source machine identifier could be added too?  I'm not sure what the intent of the source identifier shouldn't change is.  One way I'm thinking about it is that the source shouldn't change for a given producer but we could think of each machine as a new source that comes and goes.  This makes it harder for people to depend on it but gives us a lot of info on where things are coming from.

Reply:
@feanil: 

> I like the new format you're thinking about @ormsbee , do you think source sort of source machine identifier could be added too?

I'm inclined to put that in a different field, if we do. Maybe `sourcehost`, and put the hostname in there? Having it as a separate field would make it easy to log/facet/filter by host, independent of message type, or if for some reason multiple worker types are running on the same host.

> I'm not sure what the intent of the source identifier shouldn't change is. One way I'm thinking about it is that the source shouldn't change for a given producer but we could think of each machine as a new source that comes and goes. This makes it harder for people to depend on it but gives us a lot of info on where things are coming from.

I think that in the above scenario, `source` is sort of the "logical/semantic source" while `sourcehost` would be the "physical source". My hypothesis (for which I have zero data) is that having that split will make it easier to handle slight incompatibilities that might arise if other alternate services (like Richie) decide to start emitting messages that the LMS reads. If one party doesn't have much control over the other, you could imagine a bit of handling on the client side like:

```python
# The Swampcastle Catalog doesn't emit UTC dates like you're supposed to...
if source == 'swampcastle/course_catalog/web':
    # some fix here...
```

I'm a little torn, because a) it's not a great pattern to encourage; and b) I'm not sure it'll ever be an issue. But I think that it's plausible and that we do get other benefits by decoupling the logical (code) and physical (server) in reporting this information.

Reply:
I like the clarity of `sourcehost` and I think that would be useful.  For you concern about sources with incompatibilities, would some sort of schema validation/registry not resolve that? Though I do agree having a logical service this event came from is a useful piece of metadata in general.

Reply:
Finally getting back revising this. I don't think schema validation/registry will be enough to completely resolve those types of incompatibilities. In previous lives, I've seen all kinds of random incompatibilities published by vendors for what should have been the same spec like:

* Putting HTML escape codes vs. not.
* Using a literal non-breaking space character vs. a space (i.e. U+00A0 vs. U+0020).
* Encoding strings in Latin-1 vs. UTF-8.
* Sending offsets 4-byte aligned vs. not.
* Assuming indexes are 1 vs. 0-based.